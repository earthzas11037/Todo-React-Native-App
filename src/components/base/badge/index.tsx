import { badgeTv } from '@/components/base/badge/badge.style'
import { Text } from '@/components/base/text'
import React from 'react'
import { View } from 'react-native'
import { tv, type VariantProps } from 'tailwind-variants'

interface BadgeProps extends VariantProps<typeof badgeTv> {
  content?: string | number
  invisible?: boolean
  children?: React.ReactNode
  className?: string
  badgeClassName?: string
  contentClassName?: string
}

export const Badge: React.FC<BadgeProps> = ({
  children,
  content,
  size = 'medium',
  color = 'primary',
  variant = 'standard',
  invisible = false,
  className,
  badgeClassName,
  contentClassName
}) => {
  const styles = React.useMemo(() => badgeTv({ size, color, variant, invisible }), [size, color, variant, invisible])
  const displayContent =
    typeof content === 'number' && variant !== 'dot' ? (content > 99 ? '99+' : content.toString()) : content

  return (
    <View className={`${styles.container({ className })}`}>
      {children}
      {!!content && (
        <View className={`${styles.badge({ className: badgeClassName })}`}>
          <Text className={`${styles.content({ className: contentClassName })}`}>
            {variant !== 'dot' ? displayContent : ''}
          </Text>
        </View>
      )}
    </View>
  )
}
