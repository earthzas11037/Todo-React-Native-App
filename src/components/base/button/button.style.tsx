import { tv } from 'tailwind-variants'

export const buttonTv = tv({
  slots: {
    container: 'flex flex-row items-center justify-center rounded-lg px-4',
    label: 'font-notoSansThai text-base font-medium',
    indicator: 'h-6 text-white',
    icon: ''
  },
  variants: {
    size: {
      small: {
        container: 'px-4 py-2',
        label: 'text-[12px]'
      },
      medium: {
        container: 'px-5 py-3',
        label: 'text-[12px]'
      },
      large: {
        container: 'px-6 py-4',
        label: 'text-[14px]'
      }
    },
    color: {
      primary: {
        container: 'bg-primary-500',
        label: 'text-white',
        indicator: 'text-white',
        icon: 'text-white'
      },
      secondary: {
        container: 'bg-secondary-500',
        label: 'text-white',
        indicator: 'text-white',
        icon: 'text-white'
      },
      success: {
        container: 'bg-success-500',
        label: 'text-white',
        icon: 'text-white'
      },
      error: {
        container: 'bg-error-500',
        label: 'text-white',
        icon: 'text-white'
      },
      info: {
        container: 'bg-info-500',
        label: 'text-white',
        icon: 'text-white'
      },
      warning: {
        container: 'bg-warning-500',
        label: 'text-white',
        icon: 'text-white'
      }
    },
    variant: {
      contained: {
        container: '',
        label: 'text-white',
        icon: 'text-white'
      },
      outlined: {
        container: 'border bg-transparent',
        label: 'text-primary-500',
        indicator: 'text-primary-500',
        icon: 'text-primary-500'
      },
      text: {
        container: 'bg-transparent',
        label: 'text-primary-500',
        indicator: 'text-primary-500',
        icon: 'text-primary-500'
      },
      tonal: {
        container: 'bg-primary-o-main',
        label: 'text-primary-500',
        icon: 'text-primary-500'
      }
    },
    disabled: {
      true: {
        container: '',
        label: '',
        indicator: ''
      }
    },
    fullWidth: {
      true: {
        container: 'self-stretch'
      },
      false: {
        container: 'self-start'
      }
    }
  },
  compoundVariants: [
    {
      disabled: true,
      className: {
        container: 'opacity-45'
      }
    },
    // {
    //   color: 'primary',
    //   variant: ['contained', 'text'],
    //   disabled: true,
    //   className: {
    //     container: 'bg-primary-o-light',
    //     label: 'text-primary-500',
    //     icon: 'text-primary-500'
    //   }
    // },
    // {
    //   color: 'secondary',
    //   variant: ['contained', 'text'],
    //   disabled: true,
    //   className: {
    //     container: 'bg-secondary-o-main',
    //     label: 'text-secondary-900',
    //     icon: 'text-secondary-900'
    //   }
    // },
    // {
    //   color: 'success',
    //   variant: ['contained', 'text'],
    //   disabled: true,
    //   className: {
    //     container: 'bg-success-o-main',
    //     label: 'text-success-900',
    //     icon: 'text-success-900'
    //   }
    // },
    // {
    //   color: 'error',
    //   variant: ['contained', 'text'],
    //   disabled: true,
    //   className: {
    //     container: 'bg-error-o-main',
    //     label: 'text-error-900',
    //     icon: 'text-error-900'
    //   }
    // },
    // {
    //   color: 'info',
    //   variant: ['contained', 'text'],
    //   disabled: true,
    //   className: {
    //     container: 'bg-info-o-main',
    //     label: 'text-info-900',
    //     icon: 'text-info-900'
    //   }
    // },
    // {
    //   color: 'warning',
    //   variant: ['contained', 'text'],
    //   disabled: true,
    //   className: {
    //     container: 'bg-warning-o-main',
    //     label: 'text-warning-900',
    //     icon: 'text-warning-900'
    //   }
    // },
    ////
    // {
    //   color: 'primary',
    //   variant: 'outlined',
    //   disabled: true,
    //   className: {
    //     container: 'border-0 opacity-45',
    //     label: 'text-white',
    //     icon: 'text-white'
    //   }
    // },
    // {
    //   color: 'secondary',
    //   variant: 'outlined',
    //   disabled: true,
    //   className: {
    //     container: 'bg-secondary-500 border-0 opacity-45',
    //     label: 'text-white',
    //     icon: 'text-white'
    //   }
    // },
    // {
    //   color: 'success',
    //   variant: 'outlined',
    //   disabled: true,
    //   className: {
    //     container: 'bg-success-500 border-0 opacity-45',
    //     label: 'text-white',
    //     icon: 'text-white'
    //   }
    // },
    // {
    //   color: 'error',
    //   variant: 'outlined',
    //   disabled: true,
    //   className: {
    //     container: 'bg-error-500 border-0 opacity-45',
    //     label: 'text-white',
    //     icon: 'text-white'
    //   }
    // },
    // {
    //   color: 'info',
    //   variant: 'outlined',
    //   disabled: true,
    //   className: {
    //     container: 'bg-info-500 border-0 opacity-45',
    //     label: 'text-white',
    //     icon: 'text-white'
    //   }
    // },
    // {
    //   color: 'warning',
    //   variant: 'outlined',
    //   disabled: true,
    //   className: {
    //     container: 'bg-warning-500 border-0 opacity-45',
    //     label: 'text-white',
    //     icon: 'text-white'
    //   }
    // },
    ////
    {
      color: 'primary',
      variant: 'outlined',
      // disabled: false,
      className: {
        container: 'border-primary-500',
        label: 'text-primary-500',
        icon: 'text-primary-500'
      }
    },
    {
      color: 'secondary',
      variant: 'outlined',
      // disabled: false,
      className: {
        container: 'border-secondary-500',
        label: 'text-secondary-500',
        icon: 'text-secondary-500'
      }
    },
    {
      color: 'success',
      variant: 'outlined',
      // disabled: false,
      className: {
        container: 'border-success-500',
        label: 'text-success-500',
        icon: 'text-success-500'
      }
    },
    {
      color: 'error',
      variant: 'outlined',
      // disabled: false,
      className: {
        container: 'border-error-500',
        label: 'text-error-500',
        icon: 'text-error-500'
      }
    },
    {
      color: 'info',
      variant: 'outlined',
      // disabled: false,
      className: {
        container: 'border-info-500',
        label: 'text-info-500',
        icon: 'text-info-500'
      }
    },
    {
      color: 'warning',
      variant: 'outlined',
      // disabled: false,
      className: {
        container: 'border-warning-500',
        label: 'text-warning-500',
        icon: 'text-warning-500'
      }
    },
    ////
    {
      color: 'primary',
      variant: 'text',
      disabled: false,
      className: {
        container: 'border-primary-500'
      }
    },
    {
      color: 'secondary',
      variant: 'text',
      disabled: false,
      className: {
        label: 'text-secondary-500',
        icon: 'text-secondary-500'
      }
    },
    {
      color: 'success',
      variant: 'text',
      disabled: false,
      className: {
        label: 'text-success-500',
        icon: 'text-success-500'
      }
    },
    {
      color: 'error',
      variant: 'text',
      disabled: false,
      className: {
        label: 'text-error-500',
        icon: 'text-error-500'
      }
    },
    {
      color: 'info',
      variant: 'text',
      disabled: false,
      className: {
        label: 'text-info-500',
        icon: 'text-info-500'
      }
    },
    {
      color: 'warning',
      variant: 'text',
      disabled: false,
      className: {
        label: 'text-warning-500',
        icon: 'text-warning-500'
      }
    },
    ////
    {
      color: 'primary',
      variant: 'tonal',
      disabled: false,
      className: {
        container: 'bg-primary-o-main'
      }
    },
    {
      color: 'secondary',
      variant: 'tonal',
      disabled: false,
      className: {
        container: 'bg-secondary-o-main',
        label: 'text-secondary-500',
        icon: 'text-secondary-500'
      }
    },
    {
      color: 'success',
      variant: 'tonal',
      disabled: false,
      className: {
        container: 'bg-success-o-main',
        label: 'text-success-500',
        icon: 'text-success-500'
      }
    },
    {
      color: 'error',
      variant: 'tonal',
      disabled: false,
      className: {
        container: 'bg-error-o-main',
        label: 'text-error-500',
        icon: 'text-error-500'
      }
    },
    {
      color: 'info',
      variant: 'tonal',
      disabled: false,
      className: {
        container: 'bg-info-o-main',
        label: 'text-info-500',
        icon: 'text-info-500'
      }
    },
    {
      color: 'warning',
      variant: 'tonal',
      disabled: false,
      className: {
        container: 'bg-warning-o-main',
        label: 'text-warning-500',
        icon: 'text-warning-500'
      }
    }
  ],
  defaultVariants: {
    variant: 'contained',
    size: 'medium',
    color: 'primary',
    disabled: false
  }
})
