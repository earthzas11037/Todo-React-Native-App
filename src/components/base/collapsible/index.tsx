import React from 'react'
import { cssInterop } from 'nativewind'
import LibCollapsible, { CollapsibleProps as LibCollapsibleProps } from 'react-native-collapsible'

// Apply NativeWind integration
cssInterop(LibCollapsible, { className: 'style' })

interface CollapsibleProps extends LibCollapsibleProps {
  /**
   * Additional styles for the container
   */
  className?: string
  expanded: boolean
}

/**
 * Reusable Collapsible component with default props
 */
export const Collapsible: React.FC<CollapsibleProps> = ({ expanded, duration = 500, children, ...props }) => {
  return (
    <LibCollapsible collapsed={!expanded} duration={duration} {...props}>
      {children}
    </LibCollapsible>
  )
}
