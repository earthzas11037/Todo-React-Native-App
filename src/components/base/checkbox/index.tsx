import { MotiView } from 'moti'
import React, { useCallback, useEffect, useRef } from 'react'
import { Animated, I18nManager, Pressable, type PressableProps, View } from 'react-native'
import colors from '@/config/theme/colors'
import { Text } from '@/components/base/text'
import { VariantProps } from 'tailwind-variants'
import CustomIcon from '@/components/base/icon'
import { checkboxTv } from '@/components/base/checkbox/checkbox.style'

const SIZE = 18
const WIDTH = 50
const HEIGHT = 28
const THUMB_HEIGHT = 22
const THUMB_WIDTH = 22
const THUMB_OFFSET = 4

type CheckboxVariants = VariantProps<typeof checkboxTv>

export interface RootProps extends CheckboxVariants, Omit<PressableProps, 'onPress'> {
  onChange: (checked: boolean) => void
  checked?: boolean
  className?: string
  accessibilityLabel?: string
}

export type IconProps = {
  checked: boolean
  disabled?: boolean | null | undefined
  color?: 'primary' | 'secondary' | 'error' | 'info' | 'success' | 'warning'
}

export const Root = ({ checked = false, children, onChange, disabled, className = '', ...props }: RootProps) => {
  const handleChange = useCallback(() => {
    onChange(!checked)
  }, [onChange, checked])

  return (
    <Pressable
      onPress={handleChange}
      className={`flex-row items-center ${className} ${disabled ? 'opacity-50' : ''}`}
      accessibilityState={{ checked }}
      disabled={disabled}
      {...props}
    >
      {children}
    </Pressable>
  )
}

type LabelProps = {
  text: string
  className?: string
  testID?: string
}

const Label = ({ text, testID, className = '' }: LabelProps) => {
  return (
    <Text testID={testID} className={` ${className} pl-[6px]`}>
      {text}
    </Text>
  )
}

export const CheckboxIcon = ({ checked = false, color = 'primary', disabled }: IconProps) => {
  const _color = checked ? colors[color][500] : colors.text.disabled
  return (
    <MotiView
      style={{
        height: SIZE,
        width: SIZE,
        borderColor: _color,
        shadowColor: colors.primary[500],
        shadowOffset: { width: 0, height: 2 },
        shadowOpacity: checked ? 0.25 : 0,
        shadowRadius: 3.84,
        elevation: 5
      }}
      className='items-center justify-center rounded-[5px] border-2'
      from={{ backgroundColor: 'transparent', borderColor: '#CCCFD6' }}
      animate={{
        backgroundColor: checked ? _color : 'transparent',
        borderColor: _color
      }}
      transition={{
        backgroundColor: { type: 'timing', duration: 100 },
        borderColor: { type: 'timing', duration: 100 }
      }}
    >
      <MotiView
        from={{ opacity: 0 }}
        animate={{ opacity: checked ? 1 : 0 }}
        transition={{ opacity: { type: 'timing', duration: 100 } }}
      >
        <CustomIcon name='IconCheck' size={16} color='white' />
      </MotiView>
    </MotiView>
  )
}

const CheckboxRoot = ({ checked = false, children, ...props }: RootProps) => {
  return (
    <Root checked={checked} accessibilityRole='checkbox' {...props}>
      {children}
    </Root>
  )
}

const CheckboxBase = ({
  checked = false,
  testID,
  label,
  color = 'primary',
  ...props
}: RootProps & { label?: string }) => {
  return (
    <CheckboxRoot checked={checked} testID={testID} color={color} {...props}>
      <CheckboxIcon checked={checked} color={color} />
      {label ? <Label text={label} testID={testID ? `${testID}-label` : undefined} className='pr-2' /> : null}
    </CheckboxRoot>
  )
}

export const Checkbox = Object.assign(CheckboxBase, {
  Icon: CheckboxIcon,
  Root: CheckboxRoot,
  Label
})

export const RadioIcon = ({ checked = false, color = 'primary', disabled }: IconProps) => {
  const borderColor = checked ? colors[color][500] : disabled ? colors.gray.o.light : colors.text.disabled
  const innerColor = !disabled ? 'transparent' : checked ? colors[color][500] : colors.gray.o.light
  return (
    <View
      style={[
        {
          height: SIZE,
          width: SIZE,
          borderWidth: checked || (!checked && disabled) ? 5 : 2,
          justifyContent: 'center',
          alignItems: 'center',
          borderRadius: SIZE / 2,
          backgroundColor: 'transparent',
          shadowOffset: { width: 0, height: 2 },
          shadowRadius: 3.84,
          elevation: 5,
          borderColor: borderColor,
          shadowColor: disabled ? 'transparent' : colors[color][500],
          shadowOpacity: checked ? 0.25 : 0
        }
      ]}
    >
      {checked || disabled ? (
        <View
          style={[
            {
              height: 8,
              width: 8,
              borderRadius: 4
            },
            { backgroundColor: innerColor }
          ]}
        />
      ) : null}
    </View>
  )
}

const RadioRoot = ({ checked = false, children, ...props }: RootProps) => {
  return (
    <Root checked={checked} accessibilityRole='radio' {...props}>
      {children}
    </Root>
  )
}

const RadioBase = ({
  checked = false,
  testID,
  label,
  color = 'primary',
  disabled,
  ...props
}: RootProps & { label?: string }) => {
  return (
    <RadioRoot checked={checked} testID={testID} {...props}>
      <RadioIcon checked={checked} color={color} disabled={disabled} />
      {label ? <Label text={label} testID={testID ? `${testID}-label` : undefined} /> : null}
    </RadioRoot>
  )
}

export const Radio = Object.assign(RadioBase, {
  Icon: RadioIcon,
  Root: RadioRoot,
  Label
})

export const SwitchIcon = ({ checked = false }: IconProps) => {
  const translateX = useRef(new Animated.Value(checked ? THUMB_OFFSET : WIDTH - THUMB_WIDTH - THUMB_OFFSET)).current
  const backgroundColor = checked ? colors.primary[500] : colors.text.disabled

  useEffect(() => {
    Animated.timing(translateX, {
      toValue: checked ? THUMB_OFFSET : WIDTH - THUMB_WIDTH - THUMB_OFFSET,
      duration: 150,
      useNativeDriver: true
    }).start()
  }, [checked])

  return (
    <View className='w-[50px] justify-center'>
      <View className='overflow-hidden rounded-full'>
        <View
          style={{
            width: WIDTH,
            height: HEIGHT,
            backgroundColor
          }}
        />
      </View>
      <Animated.View
        style={{
          height: THUMB_HEIGHT,
          width: THUMB_WIDTH,
          position: 'absolute',
          backgroundColor: 'white',
          borderRadius: 13,
          right: 0,
          transform: [
            {
              translateX: I18nManager.isRTL ? translateX : Animated.multiply(translateX, -1)
            }
          ]
        }}
      />
    </View>
  )
}

const SwitchRoot = ({ checked = false, children, ...props }: RootProps) => {
  return (
    <Root checked={checked} accessibilityRole='switch' {...props}>
      {children}
    </Root>
  )
}

const SwitchBase = ({ checked = false, testID, label, ...props }: RootProps & { label?: string }) => {
  return (
    <SwitchRoot checked={checked} testID={testID} {...props}>
      <SwitchIcon checked={checked} />
      {label ? <Label text={label} testID={testID ? `${testID}-label` : undefined} /> : null}
    </SwitchRoot>
  )
}

export const Switch = Object.assign(SwitchBase, {
  Icon: SwitchIcon,
  Root: SwitchRoot,
  Label
})
