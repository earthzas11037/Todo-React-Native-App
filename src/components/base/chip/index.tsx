import React from 'react'
import { TouchableOpacity, Pressable } from 'react-native'
import { type VariantProps } from 'tailwind-variants'
import CustomIcon from '@/components/base/icon'
import { chipTv } from '@/components/base/chip/chip.style'
import { Text } from '@/components/base/text'

interface ChipProps extends VariantProps<typeof chipTv> {
  label: string
  icon?: JSX.Element
  deleteIcon?: JSX.Element
  onDelete?: () => void
  onPress?: () => void
  disabled?: boolean
  className?: string
  labelClassName?: string
  iconClassName?: string
  deleteIconClassName?: string
}

export const Chip: React.FC<ChipProps> = ({
  label,
  icon,
  deleteIcon,
  color = 'default',
  size = 'medium',
  disabled,
  variant = 'filled',
  fullWidth,
  onPress,
  onDelete,
  className,
  labelClassName,
  iconClassName,
  deleteIconClassName
}) => {
  const styles = React.useMemo(
    () => chipTv({ color, size, disabled, variant, fullWidth }),
    [color, size, disabled, variant, fullWidth]
  )

  return (
    <Pressable className={`${styles.container({ className })}`} onPress={onPress} disabled={disabled}>
      {icon &&
        React.cloneElement(icon as any, {
          className: `${styles.icon({ className: iconClassName })} ${icon.props.className || ''}`
        })}
      <Text className={`${styles.label({ className: labelClassName })}`}>{label}</Text>
      {onDelete && (
        <TouchableOpacity onPress={onDelete}>
          {deleteIcon ? (
            React.cloneElement(deleteIcon as any, {
              className: `${styles.deleteIcon({ className: deleteIconClassName })} ${deleteIcon.props.className || ''}`
            })
          ) : (
            <CustomIcon
              name='IconCircleX'
              size={size == 'small' ? 16 : size == 'medium' ? 20 : 24}
              className={`${styles.deleteIcon({ className: deleteIconClassName })}`}
            />
          )}
        </TouchableOpacity>
      )}
    </Pressable>
  )
}
