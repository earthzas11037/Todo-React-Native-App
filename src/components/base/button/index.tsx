import React, { JSX } from 'react'
import type { TouchableWithoutFeedbackProps } from 'react-native'
import { ActivityIndicator, TouchableWithoutFeedback, View } from 'react-native'
import type { VariantProps } from 'tailwind-variants'
import colors from '@/config/theme/colors'
import { buttonTv } from '@/components/base/button/button.style'
import { twMerge } from 'tailwind-merge'
import { LinearGradient } from '@/components/base/linear-gradient'
import { Text } from '@/components/base/text'

type ButtonVariants = VariantProps<typeof buttonTv>
interface Props extends ButtonVariants, Omit<TouchableWithoutFeedbackProps, 'disabled'> {
  label?: string
  loading?: boolean
  startIcon?: JSX.Element
  endIcon?: JSX.Element
  className?: string
  textClassName?: string
  iconClassName?: string
}

export const Button = React.forwardRef<View, Props>(
  (
    {
      label,
      loading = false,
      variant = 'contained',
      size = 'medium',
      color = 'primary',
      disabled = false,
      fullWidth = false,
      className = '',
      testID,
      textClassName = '',
      iconClassName = '',
      startIcon,
      endIcon,
      ...props
    },
    ref
  ) => {
    const [pressed, setPressed] = React.useState(false)

    const styles = React.useMemo(
      () => buttonTv({ variant, size, color, disabled, fullWidth }),
      [variant, size, color, disabled, fullWidth]
    )

    const isGradient = colors.gradient.primary && color === 'primary' && variant === 'contained' && disabled === false
    const gradientColor = colors.gradient.primary

    const handlePressIn = () => setPressed(true)
    const handlePressOut = () => setPressed(false)

    const dynamicStyle = pressed ? 'opacity-75' : ''

    if (isGradient) {
      return (
        <TouchableWithoutFeedback
          disabled={disabled || loading}
          {...props}
          testID={testID}
          onPressIn={handlePressIn}
          onPressOut={handlePressOut}
        >
          <LinearGradient
            ref={ref}
            {...(gradientColor as any)}
            className={twMerge(`${styles.container({ className })}`, dynamicStyle)}
          >
            {loading ? (
              <ActivityIndicator
                size='small'
                color='white'
                testID={testID ? `${testID}-activity-indicator` : undefined}
              />
            ) : (
              <>
                {startIcon &&
                  React.cloneElement(startIcon as any, {
                    className: `mr-[10px] ${styles.icon({ className: iconClassName })} ${
                      startIcon.props.className || ''
                    }`
                  })}
                <Text
                  testID={testID ? `${testID}-label` : undefined}
                  className={styles.label({ className: textClassName })}
                >
                  {label}
                </Text>
                {endIcon &&
                  React.cloneElement(endIcon as any, {
                    className: `ml-[10px] ${styles.icon({ className: iconClassName })} ${endIcon.props.className || ''}`
                  })}
              </>
            )}
          </LinearGradient>
        </TouchableWithoutFeedback>
      )
    }
    return (
      <TouchableWithoutFeedback
        disabled={disabled || loading}
        {...props}
        // ref={ref}
        testID={testID}
        onPressIn={handlePressIn}
        onPressOut={handlePressOut}
      >
        <View ref={ref} className={`${styles.container({ className })} ${dynamicStyle}`}>
          {loading ? (
            <ActivityIndicator
              size='small'
              className={styles.indicator()}
              testID={testID ? `${testID}-activity-indicator` : undefined}
            />
          ) : (
            <>
              {startIcon &&
                React.cloneElement(startIcon as any, {
                  className: `mr-[10px] ${styles.icon({ className: iconClassName })} ${startIcon.props.className || ''}`
                })}
              <Text
                testID={testID ? `${testID}-label` : undefined}
                className={styles.label({ className: textClassName })}
              >
                {label}
              </Text>
              {endIcon &&
                React.cloneElement(endIcon as any, {
                  className: `ml-[10px] ${styles.icon({ className: iconClassName })} ${endIcon.props.className || ''}`
                })}
            </>
          )}
        </View>
      </TouchableWithoutFeedback>
    )
  }
)
