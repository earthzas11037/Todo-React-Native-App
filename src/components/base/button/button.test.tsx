import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react-native'
import { Text, View } from 'react-native'
import { Button } from '@/components/base/button'

// Mock icon ง่ายๆ สำหรับการทดสอบ
const MockStartIcon = (props: any) => <Text {...props}>StartIcon</Text>
const MockEndIcon = (props: any) => <Text {...props}>EndIcon</Text>

describe('Button Component', () => {
  // 1. ทดสอบการแสดงผลพื้นฐาน
  it('should render the label correctly', () => {
    render(<Button label='Click Me' />)
    // ใช้ getByText เพื่อหาปุ่มจากข้อความที่ผู้ใช้เห็น
    expect(screen.getByText('Click Me')).toBeOnTheScreen()
  })

  // 2. ทดสอบการทำงานของ onPress
  it('should call onPress function when pressed', () => {
    const mockOnPress = jest.fn() // สร้าง Mock Function เพื่อติดตามการเรียกใช้
    render(<Button label='Pressable' onPress={mockOnPress} />)

    fireEvent.press(screen.getByText('Pressable'))

    expect(mockOnPress).toHaveBeenCalledTimes(1)
  })

  // 3. ทดสอบสถานะ disabled
  it('should NOT call onPress when disabled', () => {
    const mockOnPress = jest.fn()
    render(<Button label='Disabled' onPress={mockOnPress} disabled />)

    fireEvent.press(screen.getByText('Disabled'))

    expect(mockOnPress).not.toHaveBeenCalled()
  })

  // 4. ทดสอบสถานะ loading
  describe('when loading', () => {
    const mockOnPress = jest.fn()
    const testID = 'loading-button'

    beforeEach(() => {
      render(<Button label='Loading...' onPress={mockOnPress} loading testID={testID} startIcon={<MockStartIcon />} />)
    })

    it('should show an ActivityIndicator', () => {
      // ใช้ testID ที่เรากำหนดไว้เพื่อหา ActivityIndicator
      expect(screen.getByTestId(`${testID}-activity-indicator`)).toBeOnTheScreen()
    })

    it('should hide the label and icons', () => {
      // queryBy... จะ return null ถ้าไม่เจอ (ไม่ทำให้เทสต์พัง)
      expect(screen.queryByText('Loading...')).toBeNull()
      expect(screen.queryByText('StartIcon')).toBeNull()
    })

    it('should NOT call onPress when pressed', () => {
      // กดที่ตัวปุ่ม (ซึ่งตอนนี้ควรเป็น ActivityIndicator)
      fireEvent.press(screen.getByTestId(testID))
      expect(mockOnPress).not.toHaveBeenCalled()
    })
  })

  // 5. ทดสอบการแสดงผล Icon
  it('should render start and end icons', () => {
    render(<Button label='Icon Button' startIcon={<MockStartIcon />} endIcon={<MockEndIcon />} />)
    expect(screen.getByText('StartIcon')).toBeOnTheScreen()
    expect(screen.getByText('EndIcon')).toBeOnTheScreen()
  })

  // 6. ทดสอบการส่งต่อ Ref
  it('should forward ref to the underlying View', () => {
    const ref = React.createRef<View>()
    // เนื่องจาก TouchableWithoutFeedback ไม่สามารถรับ ref ได้โดยตรง
    // Component ของคุณใช้ React.forwardRef<View, Props> ซึ่งหมายถึง ref ควรไปที่ View ที่หุ้มอยู่
    render(<Button label='Ref Button' ref={ref} />)

    // ตรวจสอบว่า ref ถูกผูกกับ element จริงๆ
    expect(ref.current).toBeTruthy()
  })

  // 7. ทดสอบ Variants และ Styles ด้วย Snapshot Testing
  // วิธีนี้มีประสิทธิภาพมากในการตรวจจับการเปลี่ยนแปลงของ UI ที่ไม่คาดคิด
  describe('Variants Snapshots', () => {
    it('should render correctly with primary contained variant (Gradient)', () => {
      const { toJSON } = render(<Button label='Gradient' variant='contained' color='primary' />)
      expect(toJSON()).toMatchSnapshot()
    })

    it('should render correctly with secondary outlined variant', () => {
      const { toJSON } = render(<Button label='Outlined' variant='outlined' color='secondary' size='large' />)
      expect(toJSON()).toMatchSnapshot()
    })

    it('should render correctly with error text variant', () => {
      const { toJSON } = render(<Button label='Text' variant='text' color='error' size='small' />)
      expect(toJSON()).toMatchSnapshot()
    })

    it('should render correctly when disabled', () => {
      const { toJSON } = render(<Button label='Disabled Snap' disabled />)
      expect(toJSON()).toMatchSnapshot()
    })

    it('should render correctly with fullWidth', () => {
      const { toJSON } = render(<Button label='Full Width' fullWidth />)
      expect(toJSON()).toMatchSnapshot()
    })
  })
})
