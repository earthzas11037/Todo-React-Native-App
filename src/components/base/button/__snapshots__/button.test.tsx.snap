// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Variants Snapshots should render correctly when disabled 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  className="flex flex-row items-center justify-center rounded-lg px-5 py-3 bg-primary-500 self-start opacity-45 "
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <Text
    className="mt-[1px] font-notoSansThai font-medium text-[12px] text-white"
    style={
      {
        "writingDirection": "ltr",
      }
    }
  >
    Disabled Snap
  </Text>
</View>
`;

exports[`Button Component Variants Snapshots should render correctly with error text variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  className="flex flex-row items-center justify-center rounded-lg px-4 py-2 bg-transparent self-start "
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <Text
    className="mt-[1px] font-notoSansThai font-medium text-[12px] text-error-500"
    style={
      {
        "writingDirection": "ltr",
      }
    }
  >
    Text
  </Text>
</View>
`;

exports[`Button Component Variants Snapshots should render correctly with fullWidth 1`] = `
<ViewManagerAdapter_ExpoLinearGradient
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  colors={
    [
      4278223032,
      4281571783,
      4284920790,
    ]
  }
  endPoint={
    [
      1,
      0,
    ]
  }
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  startPoint={
    [
      0,
      0,
    ]
  }
>
  <Text
    className="mt-[1px] font-notoSansThai font-medium text-[12px] text-white"
    style={
      {
        "writingDirection": "ltr",
      }
    }
  >
    Full Width
  </Text>
</ViewManagerAdapter_ExpoLinearGradient>
`;

exports[`Button Component Variants Snapshots should render correctly with primary contained variant (Gradient) 1`] = `
<ViewManagerAdapter_ExpoLinearGradient
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  colors={
    [
      4278223032,
      4281571783,
      4284920790,
    ]
  }
  endPoint={
    [
      1,
      0,
    ]
  }
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  startPoint={
    [
      0,
      0,
    ]
  }
>
  <Text
    className="mt-[1px] font-notoSansThai font-medium text-[12px] text-white"
    style={
      {
        "writingDirection": "ltr",
      }
    }
  >
    Gradient
  </Text>
</ViewManagerAdapter_ExpoLinearGradient>
`;

exports[`Button Component Variants Snapshots should render correctly with secondary outlined variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  className="flex flex-row items-center justify-center rounded-lg px-6 py-4 border bg-transparent self-start border-secondary-500 "
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <Text
    className="mt-[1px] font-notoSansThai font-medium text-[14px] text-secondary-500"
    style={
      {
        "writingDirection": "ltr",
      }
    }
  >
    Outlined
  </Text>
</View>
`;
